Code/GraphMol/ChemReactions/SanitizeRxn.cpp

```diff
@@ -0,0 +1,349 @@
+    }
+    
+    BOOST_FOREACH(AtomInfo &pat, productAtomsToFix) {
+      if (!pat.atom) continue;
+                         
+      if(rat.bestGuessRLabel() == pat.bestGuessRLabel()) {
+        // if the atomMaps don't match, this is bad, no atomMap is ok(==0)
+        if (rat.atomMap == pat.atomMap) {
+          found = true;
+          rat.setRLabel( max_rlabel + rat.bestGuessRLabel() );
+          pat.setRLabel( max_rlabel + pat.bestGuessRLabel() );
+          if (!rat.atomMap) { // set atom mapping as well
+            rat.setAtomMap(max_atom_map + rat.bestGuessRLabel());
+            pat.setAtomMap(max_atom_map + rat.bestGuessRLabel());
+          }
+          pat.atom = NULL;
+          break;
+        }
+      }
+    }
+    
+    if(!found) {
+      throw RxnSanitizeException(makeReactantErrorMessage(
+          "Could not find RLabel mapping", rat));
+    }
+  }
+  return;
+}
+
+// if we have query atoms without rlabels, make proper rlabels if possible
+//  ensure that every rlabel in the reactant has one in the product
+
+void fixAtomMaps(ChemicalReaction &rxn) {
+  int max_atom_map = getMaxProp<int>(
+      rxn,
+      common_properties::molAtomMapNumber);
+  std::map<unsigned int, int> potential_mappings;
+
+  unsigned int templateIdx = 0;
+
+  for(MOL_SPTR_VECT::iterator it = rxn.beginReactantTemplates();
+      it != rxn.endReactantTemplates();
+      ++it, ++templateIdx) {
+    for(unsigned int idx=0;idx<(*it)->getNumAtoms(); ++idx) {
+      Atom *atom = (*it)->getAtomWithIdx(idx);
+      AtomInfo at(atom, templateIdx);
+      if(at.rlabel && !at.atomMap) {
+        if(potential_mappings.find(at.rlabel) != potential_mappings.end()) {
+            throw RxnSanitizeException(std::string("Duplicated RLabels"));
+          }
+          int map = potential_mappings[at.rlabel] = rdcast<int>(at.rlabel)+max_atom_map;
+          at.setAtomMap(map);
+      }
+    }
+  }
+
+  
+  if (!potential_mappings.size())
+    return; // everything is ok!
+
+  templateIdx = 0;
+  for(MOL_SPTR_VECT::iterator it = rxn.beginProductTemplates();
+      it != rxn.endProductTemplates();
+      ++it, ++templateIdx) {
+    for(unsigned int idx=0;idx<(*it)->getNumAtoms(); ++idx) {
+      Atom *atom = (*it)->getAtomWithIdx(idx);
+      AtomInfo at(atom, templateIdx);
+      if(at.rlabel) {
+        if(!at.atomMap) {
+          at.setAtomMap(potential_mappings[at.rlabel]);
+        } else {
+          if (at.atomMap != potential_mappings[at.rlabel]) {
+            throw RxnSanitizeException(makeProductErrorMessage(
+                "RLabel is mapped in product but not in reactant", at));
+          }
+        }
+      }
+    }
+  }
+}
+
+// might throw mol sanitization exception??? wrap in RxnSanitize?
+void fixReactantTemplateAromaticity(ChemicalReaction &rxn) {
+  unsigned int ops;
+  for(MOL_SPTR_VECT::iterator it = rxn.beginReactantTemplates();
+      it != rxn.endReactantTemplates();
+      ++it) {
+    // Cheat here, we know that this came from a RWMol
```

Comment:
Not according to the hierarchy:

```
class RWMol : public ROMol {
```

but in practice this is probably true since all ROMols need to be made somehow :)